File: backend/internal/models/warehouse.go
package models

import (
	"time"
)

// Warehouse represents a warehouse entity
type Warehouse struct {
	ID            uint       `json:"id" gorm:"primaryKey"`
	Code          string     `json:"code" gorm:"uniqueIndex;not null"`
	Name          string     `json:"name" gorm:"not null"`
	WarehouseType string     `json:"warehouse_type" gorm:"default:'main'"`
	Address       *string    `json:"address"`
	City          *string    `json:"city"`
	ManagerID     *uint      `json:"manager_id"`
	IsActive      bool       `json:"is_active" gorm:"default:true"`
	Notes         *string    `json:"notes"`
	CreatedAt     time.Time  `json:"created_at"`
	CreatedBy     *uint      `json:"created_by"`
	UpdatedAt     time.Time  `json:"updated_at"`
	UpdatedBy     *uint      `json:"updated_by"`

	// Relationships (not included in JSON by default due to omitempty)
	Manager   *User               `json:"manager,omitempty" gorm:"foreignKey:ManagerID"`
	Locations []WarehouseLocation `json:"locations,omitempty" gorm:"foreignKey:WarehouseID"`
}

// TableName specifies the table name for Warehouse
func (Warehouse) TableName() string {
	return "warehouses"
}

// SafeWarehouse represents warehouse data safe for API responses
type SafeWarehouse struct {
	ID             uint      `json:"id"`
	Code           string    `json:"code"`
	Name           string    `json:"name"`
	WarehouseType  string    `json:"warehouse_type"`
	Address        *string   `json:"address"`
	City           *string   `json:"city"`
	ManagerID      *uint     `json:"manager_id"`
	IsActive       bool      `json:"is_active"`
	Notes          *string   `json:"notes"`
	CreatedAt      time.Time `json:"created_at"`
	UpdatedAt      time.Time `json:"updated_at"`
	LocationsCount int       `json:"locations_count,omitempty"` // For list view
}

// ToSafe converts Warehouse to SafeWarehouse
func (w *Warehouse) ToSafe() *SafeWarehouse {
	safe := &SafeWarehouse{
		ID:            w.ID,
		Code:          w.Code,
		Name:          w.Name,
		WarehouseType: w.WarehouseType,
		Address:       w.Address,
		City:          w.City,
		ManagerID:     w.ManagerID,
		IsActive:      w.IsActive,
		Notes:         w.Notes,
		CreatedAt:     w.CreatedAt,
		UpdatedAt:     w.UpdatedAt,
	}

	// Count locations if loaded
	if w.Locations != nil {
		safe.LocationsCount = len(w.Locations)
	}

	return safe
}

File: frontend/src/types/warehouse.ts
export interface Warehouse {
    id: number;
    code: string;
    name: string;
    warehouse_type: string;
    address?: string;
    city?: string;
    manager_id?: number;
    is_active: boolean;
    notes?: string;
    created_at: string;
    updated_at: string;
    locations_count?: number; // For list view
    manager?: {
        id: number;
        full_name: string;
        email: string;
    };
    locations?: import('./warehouseLocation').WarehouseLocation[];
}

export interface CreateWarehouseInput {
    code: string;
    name: string;
    warehouse_type?: string;
    address?: string;
    city?: string;
    manager_id?: number;
    is_active: boolean;
    notes?: string;
}

export interface UpdateWarehouseInput {
    code?: string;
    name?: string;
    warehouse_type?: string;
    address?: string;
    city?: string;
    manager_id?: number;
    is_active?: boolean;
    notes?: string;
}

export interface WarehouseFilters {
    search?: string;
    warehouse_type?: string;
    city?: string;
    is_active?: boolean;
    page?: number;
    page_size?: number;
    sort_by?: string;
    sort_order?: 'asc' | 'desc';
}

export interface WarehouseListResponse {
    success: boolean;
    data: Warehouse[];
    pagination: {
        page: number;
        page_size: number;
        total: number;
        total_pages: number;
    };
}
